hs_core/hs_rdf.py

```diff
@@ -0,0 +1,358 @@
+            subject = s
+            break
+        if not subject:
+            raise Exception("Invalid rdf/xml, could not find required predicate dc:title")
+        return subject
+
+    def ingest_metadata(self, graph):
+        """Given an rdflib Graph, run ingest_rdf for all generic relations on the object"""
+        subject = self.rdf_subject_from_graph(graph)
+
+        generic_relations = list(filter(lambda f: isinstance(f, GenericRelation), type(self)._meta.virtual_fields))
+        for generic_relation in generic_relations:
+            generic_relation.related_model.ingest_rdf(graph, subject, self)
+        self.save()
+
+    def get_rdf_graph(self):
+        """adds the rdf triples of all generic relations on the object into an rdflib Graph"""
+        graph = Graph()
+        graph.namespace_manager = NAMESPACE_MANAGER
+
+        subject = self.rdf_subject()
+        # graph.add((subject, RDF.type, self.rdf_type()))
+        generic_relations = list(filter(lambda f: isinstance(f, GenericRelation), type(self)._meta.virtual_fields))
+        for generic_relation in generic_relations:
+            for f in getattr(self, getattr(generic_relation, 'name', None), None).all():
+                f.rdf_triples(subject, graph)
+
+        return graph
+
+    def get_xml(self, pretty_print=True, include_format_elements=True):
+        """Generates ORI+RDF xml for this metadata"""
+        # ORE = Namespace("http://www.openarchives.org/ore/terms/")
+        g = self.get_rdf_graph()
+        # g.add((self.rdf_metadata_subject(), ORE.describes, self.rdf_subject()))
+        return g.serialize(format='hydro-xml').decode()
+
+
+class RDF_Term_MixIn(object):
+    """
+    Provides methods for serializing a django model into and rdflib triples and deserializing from an rdflib Graph
+     back to a django model.  This mixin is designed for django models that are generic relations of another django
+     model which represents a group of metadata.
+
+     Without any configuration, the rdf_triples(subject) method will add all fields on the generic relation model except
+     for 'id', 'object_id', 'content_type' will be serialized as rdf triples.  If rdf_term is defined, it will be
+     used as the predicate.  If rdf_term is not defined, one will be defined using the class name and the
+     HSTERMS namespace.
+
+     When a model requires an rdf-xml format that the default configuration will not provide, you may overwrite
+     the rdf_triples method and the ingest_rdf method.  If you overwrite one, you must overwrite both or this mixin
+     will not work correctly.
+     """
+
+    ignored_fields = ['id', 'object_id', 'content_type']
+
+    def rdf_triples(self, subject, graph):
+        """Default implementation that parses by convention."""
+        term = self.get_class_term()
+        if not term:
+            metadata_node = subject
+        else:
+            metadata_node = BNode()
+            graph.add((subject, term, metadata_node))
```

Comment:
What happens if subject is empty? I think it'd be good to add a check here to only add subject to graph when it is not empty.

Reply:
I'm not seeing what we would gain from that.  if subject were empty, than we'd have to raise an exception.  An exception would be raised by the graph is subject was empty already.  I don't expect subject to ever be empty as a subject comes from the rdf_subject method which by default raises a NotImplementedError.  A subject could also be a BNODE from a parent triple.